src/document/actions/index.ts

```diff
@@ -6,96 +7,121 @@ import {
+
+        if (isLatestOpNOOP) {
+            draft.skip += lastOperation.skip;
+            draft.document.operations[scope].pop();
+        }
+
+        if (draft.document.operations[scope].length < draft.skip) {
+            throw new Error(`Cannot undo: you can't undo more operations than the ones in the scope history`);
+        }
+
+        const operationsLastIndex = draft.document.operations[scope].length - 1;
+        let skippedOpsLeft = (draft.action as UndoAction).input;
+        let index = isLatestOpNOOP ? operationsLastIndex - lastOperation.skip : operationsLastIndex;
+
+        while (skippedOpsLeft > 0 && index >= 0) {
+            const op = draft.document.operations[scope][index];
+
+            if (op.type === 'NOOP' && op.skip > 0) {
+                index = index - (op.skip + 1);
+                draft.skip += op.skip + 1;
+            } else {
+                draft.document.clipboard.push({ ...op });
+                skippedOpsLeft--;
+                index--;
+            }
+        }
+        
+        draft.action = noop(scope);
+    });
 }
 
-// redoes the last `count` undone operations
 export function redoOperation<T, A extends Action, L>(
     document: Document<T, A, L>,
     action: RedoAction,
-    wrappedReducer: ImmutableStateReducer<T, A, L>,
-): Document<T, A, L> {
-    const { scope, input: count } = action;
-    // the number of undone operations is retrieved from the revision number
-    const undoCount =
-        document.operations[scope].length - document.revision[scope];
-    if (!undoCount) {
-        throw new Error('There is no UNDO operation to REDO');
-    }
-
-    // redo can't be higher than the number of undone operations
-    const redoCount = count < undoCount ? count : undoCount;
-
-    // builds state from the initial date taking
-    // into account the redone operations
-    const operations = document.operations[scope].slice(
-        0,
-        document.revision[scope] + redoCount,
-    );
-    const newDocument = replayOperations(
-        document.initialState,
-        {
-            ...document.operations,
-            [scope]: operations,
-        },
-        wrappedReducer,
-    );
-
-    // updates the state and the revision number but
-    // keeps the operations history to allow more REDOs
-    return {
-        ...newDocument,
-        operations: document.operations,
-        revision: {
-            ...document.revision,
-            [scope]: document.revision[scope] + redoCount,
-        },
+    skip: number,
+): UndoRedoProcessResult<T, A, L> {
+    const scope = action.scope;
+
+    const defaultResult: UndoRedoProcessResult<T, A, L> = {
+        document,
+        action,
+        skip,
     };
+
+    return produce(defaultResult, draft => {
+        if (draft.skip > 0) {
+            throw new Error(`Cannot redo: skip value from reducer cannot be used with REDO action`);
+        }
+
+        if ((draft.action as RedoAction).input > 1) {
```

Comment:
Same as above

Reply:
fixed in [b39596d](https://github.com/powerhouse-inc/document-model/pull/29/commits/b39596da1c0fad4c742170a6c43134c4799d830f)